main/src/ca/uwaterloo/flix/language/phase/optimizer/LambdaDrop.scala

```diff
@@ -0,0 +1,407 @@
+
+import scala.collection.mutable
+
+/**
+  * Rewrites recursive higher-order functions to non-recursive functions with
+  * a recursive local def where constant parameters become closure captured.
+  *
+  * A constant parameter is one that has the same value in every recursive call.
+  *
+  * Example:
+  *
+  * {{{
+  *   def map(f: a -> b, l: List[a]): List[b] = match l {
+  *       case Nil     => Nil
+  *       case x :: xs => f(x) :: map(f, xs)
+  *   }
+  * }}}
+  *
+  * Is rewritten to:
+  *
+  * {{{
+  *   def map(f: a -> b, l: List[a]): List[b] =
+  *       def map$Loop(l1) = match l1 {
+  *           case Nil      => Nil
+  *           case x :: xs  => f(x) :: map$Loop(xs)
+  *       };
+  *       map$Loop(l)
+  * }}}
+  */
+object LambdaDrop {
+
+  def run(root: MonoAst.Root)(implicit flix: Flix): MonoAst.Root = flix.phase("LambdaDrop") {
+    val newDefs = ParOps.parMapValues(root.defs)(visitDef)
+    root.copy(defs = newDefs)
+  }
+
+  private def visitDef(defn: MonoAst.Def)(implicit flix: Flix): MonoAst.Def = {
+    implicit val ctx: LocalContext = LocalContext.mk()
+    if (isRewritable(defn))
+      rewriteDefn(defn)
+    else
+      defn
+  }
+
+  private def isRewritable(defn: MonoAst.Def)(implicit ctx: LocalContext): Boolean = {
+    if (isHigherOrder(defn.spec.fparams)) { // Only visit exp if it is higher-order
+      visitExp(defn.exp)(defn.sym, ctx)
+      val containsRecursiveCall = ctx.recursiveCalls.nonEmpty
+      containsRecursiveCall
+    } else {
+      false
+    }
+  }
+
+  private def isHigherOrder(fparams: List[MonoAst.FormalParam]): Boolean = {
+    fparams.exists {
+      fp =>
+        fp.tpe.typeConstructor match {
+          case Some(TypeConstructor.Arrow(_)) => true
+          case Some(TypeConstructor.ArrowWithoutEffect(_)) => true
+          case Some(_) => true
+          case None => false
+        }
+    }
+  }
+
+  private def rewriteDefn(defn: MonoAst.Def)(implicit ctx: LocalContext, flix: Flix): MonoAst.Def = {
+    implicit val params: List[(MonoAst.FormalParam, ParamKind)] = paramKinds(ctx.recursiveCalls, defn.spec.fparams)
+    implicit val (newDefnSym, subst): (Symbol.VarSym, Substitution) = mkSubst(defn, params)
+    val rewrittenExp = rewriteExp(defn.exp)(defn.sym, newDefnSym, subst, params)
+    val body = mkLocalDefExpr(rewrittenExp, newDefnSym)
+    defn.copy(exp = body)
+  }
+
+  private def visitExp(exp0: MonoAst.Expr)(implicit sym0: Symbol.DefnSym, ctx: LocalContext): Unit = exp0 match {
+    case Expr.Cst(_, _, _) =>
+
+    case Expr.Var(_, _, _) =>
+
+    case Expr.Lambda(_, exp, _, _) =>
+      visitExp(exp)
+
+    case Expr.ApplyAtomic(_, exps, _, _, _) =>
+      exps.foreach(visitExp)
+
+    case Expr.ApplyClo(exp1, exp2, _, _, _) =>
+      visitExp(exp1)
+      visitExp(exp2)
+
+    case Expr.ApplyDef(sym, exps, _, _, _, _) =>
+      // Check for recursion
+      if (sym == sym0) {
+        val applyExp = exp0.asInstanceOf[Expr.ApplyDef]
```

Comment:
instead you can bind with `expr@Expr.ApplyDef`

Reply:
You have previously said no to this, hence the cast ;)